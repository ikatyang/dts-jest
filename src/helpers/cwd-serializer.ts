export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
