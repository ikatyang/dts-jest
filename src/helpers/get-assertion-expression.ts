import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
