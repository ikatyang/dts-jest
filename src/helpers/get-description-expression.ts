import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
