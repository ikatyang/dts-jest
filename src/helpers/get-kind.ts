import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
