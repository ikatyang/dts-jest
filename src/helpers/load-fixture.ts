import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
