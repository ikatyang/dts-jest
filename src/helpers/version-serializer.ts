const version_regex = /\bv[0-9]+\.[0-9]+\.[0-9]+\b/g;

export const version_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(version_regex, 'vX.Y.Z-mocked')),
  test: (value: any) => typeof value === 'string' && version_regex.test(value),
};
