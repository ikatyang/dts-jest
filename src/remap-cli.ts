import * as fs from 'fs';
import globby = require('globby');
import intersection = require('lodash.intersection');
import make_dir = require('make-dir');
import * as path from 'path';
import yargs = require('yargs');
import { package_homepage, package_remap_bin } from './definitions';
import { remap } from './remap';
import { get_typescript } from './utils/get-typescript';

export interface RemapCliOptions {
  check?: boolean;
  listDiff?: boolean;
  outDir?: string;
  rename?: string;
  typescript?: string;
}

export interface RemapCliArgs extends RemapCliOptions {
  _: string[];
}

const remap_cli_parser = get_argv_parser();

export const remap_cli = (argv: string[]): void => {
  // istanbul ignore next
  const {
    _: globs,
    check = false,
    listDiff: list_diff = false,
    outDir: output_directory,
    rename: rename_template,
    typescript: typescript_id,
  } = remap_cli_parser.parse(argv);

  const filepaths = globby.sync(globs, { absolute: true });

  if (filepaths.length === 0) {
    log('There is no matched file.');
    return;
  }

  const invalid_filepaths = get_invalid_filepaths(
    filepaths,
    output_directory,
    rename_template,
  );

  if (invalid_filepaths.length !== 0) {
    throw new Error(
      `Input and output filename cannot be the same:\n\n${invalid_filepaths
        .map(x => `  ${x}`)
        .join('\n')}\n\nConsider adjusting option: --outDir or --rename`,
    );
  }

  if (!check && output_directory !== undefined) {
    make_dir.sync(output_directory);
  }

  const { ts, ts_path } = get_typescript(typescript_id, '');

  log(`Using TypeScript v${ts.version} from ${ts_path}`);

  let is_check_success = true;

  filepaths.forEach(filepath => {
    const source_content = fs.readFileSync(filepath, 'utf8');

    const snapshot_filepath = get_snapshot_filepath(filepath);
    const snapshot_content = fs.readFileSync(snapshot_filepath, 'utf8');

    const output_filepath = get_output_filepath(
      filepath,
      output_directory,
      rename_template,
    );
    const output_content = remap(source_content, snapshot_content, {
      typescript: ts,
      source_filename: filepath,
    });

    if (check || list_diff) {
      const target_content = fs.readFileSync(output_filepath, 'utf8');

      if (output_content !== target_content) {
        if (check) {
          is_check_success = false;
        }

        if (list_diff) {
          log(output_filepath);
        }
      }
    }

    if (!check) {
      fs.writeFileSync(output_filepath, output_content);
    }
  });

  if (!is_check_success) {
    throw new Error(`Difference(s) detected`);
  }
};

function get_invalid_filepaths(
  filepaths: string[],
  output_directory?: string,
  rename_template?: string,
) {
  return intersection(
    filepaths,
    filepaths.map(filepath =>
      get_output_filepath(filepath, output_directory, rename_template),
    ),
  );
}

function get_output_filepath(
  filepath: string,
  output_directory?: string,
  rename_template?: string,
) {
  const redirect_dirpath =
    output_directory === undefined
      ? path.dirname(filepath)
      : path.resolve(output_directory);

  const renamed_filename = rename_filename(filepath, rename_template);

  return path.join(redirect_dirpath, renamed_filename);
}

function rename_filename(filepath: string, template?: string) {
  const filename = path.basename(filepath);

  if (template === undefined) {
    return filename;
  }

  const extname = path.extname(filepath);
  const basename = path.basename(filepath, extname);

  return template
    .replace('{{filename}}', filename)
    .replace('{{basename}}', basename)
    .replace('{{extname}}', extname.slice(1) /* remove leading dot */);
}

function get_snapshot_filepath(filepath: string) {
  return path.join(
    path.dirname(filepath),
    '__snapshots__',
    `${path.basename(filepath)}.snap`,
  );
}

function get_argv_parser() {
  let parser = yargs
    .usage(
      `Usage: ${package_remap_bin} [--outDir <path>] [--rename <template>] <TestFileGlob> ...`,
    )
    .demandCommand(1)
    .epilogue(`Documentation: ${package_homepage}`);

  const options = get_cli_options();

  ['help', 'version', ...Object.keys(options)].sort().forEach(key => {
    switch (key) {
      case 'help':
      case 'version':
        parser = parser[key]().alias(key, key[0]);
        break;
      default:
        parser = parser.option(key, options[key as keyof RemapCliOptions]);
        break;
    }
  });

  return parser;
}

function get_cli_options(): {
  [K in keyof RemapCliOptions]: {
    type: 'boolean' | 'string';
    alias: string;
    description: string;
  }
} {
  return {
    check: {
      alias: 'c',
      type: 'boolean',
      description: 'Throw error if target is different with output',
    },
    listDiff: {
      alias: 'l',
      type: 'boolean',
      description: 'Print filename if target is different with output',
    },
    outDir: {
      alias: 'o',
      type: 'string',
      description: 'Redirect output structure to the directory',
    },
    rename: {
      alias: 'r',
      type: 'string',
      description:
        'Rename output filename using template {{variable}}, available variable: filename, basename, extname',
    },
    typescript: {
      alias: 't',
      type: 'string',
      description: 'Specify which TypeScript source to use',
    },
  };
}

function log(message: string) {
  // tslint:disable-next-line:no-console
  console.log(message);
}
