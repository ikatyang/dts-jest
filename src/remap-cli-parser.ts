import yargs = require('yargs');
import { package_homepage, package_remap_bin } from './definitions';
import { RemapCliArguments, RemapCliOptions } from './remap-cli';

export const create_remap_cli_parser = () => {
  let parser = yargs
    .usage(
      `Usage: ${package_remap_bin} [--outDir <path>] [--rename <template>] <TestFileGlob> ...`,
    )
    .demandCommand(1)
    .epilogue(`Documentation: ${package_homepage}`);

  const options = get_cli_options();
  const special_options = {
    help: { alias: 'h' },
    version: { alias: 'v' },
  };

  const is_special_key = (key: string): key is keyof typeof special_options =>
    key in special_options;

  [...Object.keys(options), ...Object.keys(special_options)]
    .sort()
    .forEach(key => {
      parser = is_special_key(key)
        ? parser[key]().alias(key, special_options[key].alias)
        : parser.option(key, options[key as keyof RemapCliOptions]);
    });

  return parser.parse.bind(parser) as (argv: string[]) => RemapCliArguments;
};

function get_cli_options(): {
  [K in keyof RemapCliOptions]: {
    type: 'boolean' | 'string';
    alias: string;
    description: string;
  }
} {
  return {
    check: {
      alias: 'c',
      type: 'boolean',
      description:
        'Throw error if target content is different from output content',
    },
    listDifferent: {
      alias: 'l',
      type: 'boolean',
      description:
        'Print the filenames of files that their target content is different from output content',
    },
    outDir: {
      alias: 'o',
      type: 'string',
      description: 'Redirect output structure to the directory',
    },
    rename: {
      alias: 'r',
      type: 'string',
      description:
        'Rename output filename using template {{variable}}, available variable: filename, basename, extname',
    },
    typescript: {
      alias: 't',
      type: 'string',
      description: 'Specify which TypeScript source to use',
    },
  };
}
