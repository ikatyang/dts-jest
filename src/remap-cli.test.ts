import * as fs from 'fs';
import { cwd_serializer } from './helpers/cwd-serializer';
import { get_fixture_filename } from './helpers/load-fixture';
import { version_serializer } from './helpers/version-serializer';
import { remap_cli } from './remap-cli';
import { create_remap_cli_parser } from './remap-cli-parser';

expect.addSnapshotSerializer(cwd_serializer);
expect.addSnapshotSerializer(version_serializer);

const console_log = jest.spyOn(console, 'log');
const console_error = jest.spyOn(console, 'error');
const wrtie_file_sync = jest.spyOn(fs, 'writeFileSync');
const process_exit = jest.spyOn(process, 'exit');

beforeAll(() => {
  console_log.mockReturnThis();
  console_error.mockReturnThis();
  wrtie_file_sync.mockReturnThis();
  process_exit.mockImplementation(code => {
    if (code !== 0) {
      throw new Error(`Exit with code ${code}`);
    }
  });
});

beforeEach(() => {
  console_log.mockClear();
  console_error.mockClear();
  wrtie_file_sync.mockClear();
});

afterAll(() => {
  console_log.mockRestore();
  console_error.mockRestore();
  wrtie_file_sync.mockRestore();
  process_exit.mockRestore();
});

const remap_cli_parser = create_remap_cli_parser();

it('should remap correctly', () => {
  run_cli('example.ts', ['--rename', '{{basename}}.snap.{{extname}}']);
  expect(wrtie_file_sync.mock.calls[0][1]).toMatchSnapshot();
});

it('should write path correctly with --rename', () => {
  run_cli('example.ts', ['--rename', '{{basename}}.snap.{{extname}}']);
  expect(wrtie_file_sync.mock.calls[0][0]).toMatchSnapshot();
});

it('should write path correctly with --outDir', () => {
  run_cli('example.ts', ['--outDir', './snapshots']);
  expect(wrtie_file_sync.mock.calls[0][0]).toMatchSnapshot();
});

it('should log warning if there is no matched file', () => {
  run_cli('---not-exist---', []);
  expect(get_console_content(console_log)).toMatchSnapshot();
});

it('should log diff filenames correctly with --listDifferent', () => {
  run_cli('example.ts', ['--listDifferent', '--rename', 'empty.ts']);
  expect(get_console_content(console_log)).toMatchSnapshot();
});

it('should log diff filenames correctly with --listDifferent if there is no such output file', () => {
  run_cli('example.ts', ['--listDifferent', '--rename', '___not_exist___.ts']);
  expect(get_console_content(console_log)).toMatchSnapshot();
});

it('should throw error if differences exist with --check', () => {
  expect(() =>
    run_cli('example.ts', ['--check', '--rename', 'empty.ts']),
  ).toThrowErrorMatchingSnapshot();
});

it('should not throw error if there is no difference with --check', () => {
  expect(() =>
    run_cli('example.ts', ['--check', '--rename', 'remapped.ts']),
  ).not.toThrow();
});

it('should throw error with differences if differences exist with --check and --listDifferent', () => {
  expect(() =>
    run_cli('example.ts', [
      '--check',
      '--listDifferent',
      '--rename',
      'empty.ts',
    ]),
  ).toThrowErrorMatchingSnapshot();
});

it('should throw error if there is no input', () => {
  expect(() => remap_cli(remap_cli_parser([]))).toThrow();
  expect(get_console_content(console_error)).toMatchSnapshot();
});

it('should throw error if input and output file are the same', () => {
  expect(() => run_cli('example.ts', [])).toThrowErrorMatchingSnapshot();
});

it('should log typescript information', () => {
  run_cli('example.ts', ['--rename', '{{basename}}.snap.{{extname}}']);
  expect(get_console_content(console_log)).toMatchSnapshot();
});

function run_cli(test_glob: string, args: string[]) {
  const filename = get_fixture_filename(`remap-cli/${test_glob}`);
  remap_cli(remap_cli_parser([filename, ...args]));
}

function get_console_content(console_fn: jest.SpyInstance<typeof console.log>) {
  return console_fn.mock.calls.map(x => x[0]).join('\n');
}
