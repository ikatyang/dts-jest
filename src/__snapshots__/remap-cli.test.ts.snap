// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should log diff filenames correctly with --listDifferent 1`] = `
"Using TypeScript vX.Y.Z-mocked from <cwd>/node_modules/typescript/lib/typescript.js

Listing files that their target content is different from output content:

  <cwd>/fixtures/remap-cli/example.ts -> <cwd>/fixtures/remap-cli/empty.ts"
`;

exports[`should log diff filenames correctly with --listDifferent if there is no such output file 1`] = `
"Using TypeScript vX.Y.Z-mocked from <cwd>/node_modules/typescript/lib/typescript.js

Listing files that their target content is different from output content:

  <cwd>/fixtures/remap-cli/example.ts -> <cwd>/fixtures/remap-cli/___not_exist___.ts"
`;

exports[`should log typescript information 1`] = `"Using TypeScript vX.Y.Z-mocked from <cwd>/node_modules/typescript/lib/typescript.js"`;

exports[`should log warning if there is no matched file 1`] = `"There is no matched file."`;

exports[`should remap correctly 1`] = `
"// @dts-jest:snap -> (mocked max)
Math.max(3, 2, 1);

// @dts-jest:snap description-min -> (mocked min)
Math.min(3, 2, 1);

// @dts-jest:group A

// @dts-jest:snap -> (mocked abs)
Math.abs(3);

// @dts-jest:snap description-sin -> (mocked sin)
Math.sin(0);

// @dts-jest:group B

// @dts-jest:snap -> (mocked cos)
Math.cos(0);

// @dts-jest:snap -> (mocked tan 1)
Math.tan(0);

// @dts-jest:snap -> (mocked tan 2)
Math.tan(0);
"
`;

exports[`should throw error if differences exist with --check 1`] = `"Difference(s) detected"`;

exports[`should throw error if input and output file are the same 1`] = `
"Input and output filename cannot be the same:

  <cwd>/fixtures/remap-cli/example.ts

Consider adjusting option \`--outDir\` or \`--rename\` to redirect output"
`;

exports[`should throw error if there is no input 1`] = `
"Usage: dts-jest-remap [--outDir <path>] [--rename <template>] <TestFileGlob> ...

Options:
  --check, -c          Throw error if target content is different from output
                       content                                         [boolean]
  --help, -h           Show help                                       [boolean]
  --listDifferent, -l  Print the filenames of files that their target content is
                       different from output content                   [boolean]
  --outDir, -o         Redirect output structure to the directory       [string]
  --rename, -r         Rename output filename using template {{variable}},
                       available variable: filename, basename, extname  [string]
  --typescript, -t     Specify which TypeScript source to use           [string]
  --version, -v        Show version number                             [boolean]

Documentation: https://github.com/ikatyang/dts-jest#readme

Not enough non-option arguments: got 0, need at least 1"
`;

exports[`should throw error with differences if differences exist with --check and --listDifferent 1`] = `
"Listing files that their target content is different from output content:

  <cwd>/fixtures/remap-cli/example.ts -> <cwd>/fixtures/remap-cli/empty.ts

Difference(s) detected"
`;

exports[`should write path correctly with --outDir 1`] = `"<cwd>/snapshots/example.ts"`;

exports[`should write path correctly with --rename 1`] = `"<cwd>/fixtures/remap-cli/example.snap.ts"`;
